### ISSUE TITLE
Trigger Rule based on all ancestors

### ISSUE THREAD

**Issue by jayesh-shaw-glean** [apache/airflow#33096]
### Description

Currently Airflow supports a number of trigger_rules all of which cater to various use cases. But notably all of the trigger rules currently depend only on direct parents. It would be nice to have trigger rules that let a task depend on all of its ancestors and not only direct parents.

### Use case/motivation

For an example take the following dag:
```
PRE_PROCESS >> A >>  B >> C
```

This dag takes a list of tasks it needs to run and skips the tasks that do not need to run. This is achieved by utilising the PRE_PROCESS node, it takes the list of tasks that need to be run and then gets the list of all tasks in the dag, with which it constructs the list of tasks that do not need to run, and then leverages the `SkipMixin.skip()` function to achieve the goal. 

The issue: Consider a scenario where we need to run only A and C. In this case, the PRE_PROCESS operator would skip B and the dag state would look like the following:

```
PRE_PROCESS (success) >> A (queued) >> B (skipped) >> C (queued)
```

In current set of trigger rules, none of them allow us to make sure that C runs only after A succeeds. C's trigger could only depend on B's status. For example, if the operators have the `none_failed` trigger rule, the state of the dag would become:

```
PRE_PROCESS (success) >> A (running) >> B (skipped) >> C (running)
```

It would thus be nice to have a trigger_rule which caters the above use case.

### Current workaround:

Our use-case involves a DAG with ~200 tasks and at times we might only need to run 2-4 operations (which may or may not depend on each other directly/indirectly). To solve the problem above, what we do is
   - Set the trigger_rule of all tasks to `none_failed`
   - Communicate the list of tasks that need to be run to all tasks via XCom
   - In a task determine if it needs to be run, if not raise `AirflowSkipException`
   
The above solution works well, but the issue arises when we attempt to run on scale (say have 100-200 concurrent runs of the same dag, each running only 2-4 operations). In such cases, every operation that needs to be skipped consumes a worker slot for a little while and on scale, this slows the dag by a pretty good fraction and moreover this is technically overusage of resources.

Happy to raise a PR for the same, but wanted to know if this problem has been discussed previously and since I am unfamiliar with Airflow's source code, would love to know the feasibility of implementing this feature from people who are familiar with the stack. 

Thank you!

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
It's an incredibly complex and extremely difficult task for performance reasons. I seriously doubt you will be able to implement it in the way you want that will handle 1000s or even 100s tasks.

**Comment by jayesh-shaw-glean** [comment]
> It's an incredibly complex and extremely difficult task for performance reasons. I seriously doubt you will be able to implement it in the way you want that will handle 1000s or even 100s tasks.

Yeah, sounds fair. We'll be trading usage of resources from one place to another, which could possibly be costlier. Implementing such a trigger rule is basically adding implicit edges between tasks and the number of edges in the dag could go up 10-20 in our use-case

**Comment by potiuk** [comment]
This is more than that. When you implement your "custom" solution, you are limited to what you use, but if you want to generalize as "product feature" , well the cases might be much more complex. in your case it might be 10-20 but there are users who have 10s of thousands tasks in their DAGs.

**Comment by eladkal** [comment]
Given
`X1 >> X2 >> X3 >> X4`

My argument here is: if X4 has a question about the status of X1 then X4 should have direct dependency on X1.
Thus having also `X1 >> X4`.
Alternatively, you can write an operator that checks upstream tasks statuses and generate logic around it as you please.

**Comment by eladkal** [comment]
I'm converting this into a discussion so we can continue talk about it.
If/When we will have solid idea of how to address this we can generate issue with the information about the required task.
